ecs-agent/acs/session/refresh_credentials_responder.go

```diff
@@ -0,0 +1,199 @@
+	credsMetadataSetter CredentialsMetadataSetter,
+	metricsFactory metrics.EntryFactory,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &refreshCredentialsResponder{
+		credentialsManager:  credentialsManager,
+		credsMetadataSetter: credsMetadataSetter,
+		metricsFactory:      metricsFactory,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*refreshCredentialsResponder) Name() string { return "refresh credentials responder" }
+
+func (r *refreshCredentialsResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handleCredentialsMessage
+}
+
+func (r *refreshCredentialsResponder) handleCredentialsMessage(message *ecsacs.IAMRoleCredentialsMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", RefreshCredentialsMessageName))
+	messageID := aws.StringValue(message.MessageId)
+	taskARN := aws.StringValue(message.TaskArn)
+	metricFields := logger.Fields{
+		field.MessageID: messageID,
+		field.TaskARN:   taskARN,
+	}
+
+	// Validate fields in the message.
+	err := validateIAMRoleCredentialsMessage(message)
+	if err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", RefreshCredentialsMessageName),
+			logger.Fields{
+				field.Error: err,
+			})
+		err = errors.Wrap(err, "ACS refresh credentials message validation failed")
+		r.metricsFactory.New(metrics.CredentialsRefreshFailure).WithFields(metricFields).Done(err)
+		return
+	}
+
+	// Handle credentials refresh.
+	err = r.credentialsManager.SetTaskCredentials(&credentials.TaskIAMRoleCredentials{
+		ARN: taskARN,
+		IAMRoleCredentials: credentials.IAMRoleCredentialsFromACS(message.RoleCredentials,
+			aws.StringValue(message.RoleType)),
+	})
+	if err != nil {
+		logger.Error(fmt.Sprintf("Unable to handle %s due to error in setting credentials",
+			RefreshCredentialsMessageName), logger.Fields{
+			field.MessageID: messageID,
+			field.Error:     err,
+		})
+		err = errors.Wrap(err, "unable to set credentials in the credentials manager")
+		r.metricsFactory.New(metrics.CredentialsRefreshFailure).WithFields(metricFields).Done(err)
+		return
+	}
+	err = r.setCredentialsMetadata(message)
+	if err != nil {
+		logger.Error(fmt.Sprintf("Unable to handle %s due to error in setting credentials metadata",
+			RefreshCredentialsMessageName), logger.Fields{
+			field.MessageID: messageID,
+			field.Error:     err,
+		})
+		r.metricsFactory.New(metrics.CredentialsRefreshFailure).WithFields(metricFields).Done(err)
+		return
+	}
+
+	// Send ACK.
+	err = r.respond(&ecsacs.IAMRoleCredentialsAckRequest{
+		Expiration:    message.RoleCredentials.Expiration,
+		MessageId:     message.MessageId,
+		CredentialsId: message.RoleCredentials.CredentialsId,
+	})
+	if err != nil {
+		logger.Warn(fmt.Sprintf("Error acknowledging %s", RefreshCredentialsMessageName), logger.Fields{
+			field.MessageID: messageID,
+			field.Error:     err,
+		})
+		err = errors.Wrapf(err, "unable to ACK task credentials for task with ARN %s", taskARN)
```

Comment:
Can you update logging here too to something like `unable to ACK task credentials for RefreshCredentialsMessage for task ARN` etc.?

Reply:
Or `unable to ACK refreshCredentialsMessage`?

Reply:
Chatted offline - log context should provides enough information.

Reply:
Further context - L123 displays the message type in the logging statement https://github.com/aws/amazon-ecs-agent/pull/3830/files#diff-54706336185a9c70284090b1f5e17ea6916c57c272e5de8fe3db7fe819e7a4cdR123

(i.e., this results in log text `Error acknowledging IAMRoleCredentialsMessage`)